### PR TITLE
Replace occurences of alias_method_chain with their Module#prepend counterpart

### PR DESCRIPTION
**Pull request by kirs**
Earlier @fbernier proposed a cool idea of replacing `alias_method_chain` with `Module#prepend` in #17394.
I guess it's now time to merge it :rocket:
/cc @rafaelfranca

### CODE REVIEW

**Comment by fbernier** [comment]
Thanks for this, looks like some tests are failing. I will be putting some more work on it.

File: activesupport/lib/active_support/core_ext/marshal.rb [hunk 1]
```diff
@@ -1,9 +1,6 @@
-require 'active_support/core_ext/module/aliasing'
-
-module Marshal
-  class << self
-    def load_with_autoloading(source)
-      load_without_autoloading(source)
+module MarshalWithAutoloading
+    def load(source)
```
Comment (kaspth) [inline activesupport/lib/active_support/core_ext/marshal.rb]: Outdent this by two spaces

File: activesupport/lib/active_support/core_ext/marshal.rb [hunk 2]
```diff
@@ -15,7 +12,10 @@ def load_with_autoloading(source)
         raise exc
       end
     end
+end
 
-    alias_method_chain :load, :autoloading
+Marshal.module_eval do
```
Comment (kaspth) [inline activesupport/lib/active_support/core_ext/marshal.rb]: Would using class eval remove the need to use "<< self"

File: activesupport/lib/active_support/core_ext/range/include_range.rb [hunk 3]
```diff
@@ -9,15 +7,15 @@ class Range
...
-  def include_with_range?(value)
+  def include?(value)
     if value.is_a?(::Range)
       # 1...10 includes 1..9 but it does not include 1..10.
       operator = exclude_end? && !value.exclude_end? ? :< : :<=
-      include_without_range?(value.first) && value.last.send(operator, last)
+      super(value.first) && value.last.send(operator, last)
     else
-      include_without_range?(value)
+      super(value)
```
Comment (kaspth) [inline activesupport/lib/active_support/core_ext/range/include_range.rb]: No need to pass the arg, `super` will do.

File: activesupport/test/core_ext/marshal_test.rb [hunk 4]
```diff
@@ -15,7 +15,7 @@ def teardown
     sanity_data = ["test", [1, 2, 3], {a: [1, 2, 3]}, ActiveSupport::TestCase]
     sanity_data.each do |obj|
       dumped = Marshal.dump(obj)
-      assert_equal Marshal.load_without_autoloading(dumped), Marshal.load(dumped)
+      assert_nothing_raised { Marshal.load(dumped) }
```
Comment (kaspth) [inline activesupport/test/core_ext/marshal_test.rb]: I'm not sure this is the right change, but I got no suggestions right now.

**Comment by kaspth** [comment]
Should we follow through with a deprecation on `alias_method_chain`?

File: activesupport/lib/active_support/core_ext/marshal.rb (hunk 2 above)
Comment (sgrif) [inline activesupport/lib/active_support/core_ext/marshal.rb]: `prepend` is public in 2.2, you can just call `Marshal.prepend` directly.

Comment (sgrif) [inline activesupport/lib/active_support/core_ext/marshal.rb]: Ah nevermind I see what you're doing

Comment (sgrif) [inline activesupport/lib/active_support/core_ext/marshal.rb]: How about `Marshal.singleton_class.prepend(MarshalWithAutoloading)`?

Comment (kaspth) [inline activesupport/lib/active_support/core_ext/marshal.rb]: :+1: for Sean's suggestion.

**Comment by kirs** [comment]
Thanks for review! I've pushed some updated parts.

@kaspth it would be great to deprecate `alias_method_chain`, but we need first to update the JSON encoding too: https://github.com/rails/rails/blob/master/activesupport/lib/active_support/core_ext/object/json.rb#L41

I tried it and I also tried to fix the failing test, but I had no success yet. I will spend some more time on it tomorrow :yellow_heart:

**Comment by kaspth** [comment]
Well, the question was more aimed at any Core Team members that might have happened to listen in. But yes, do look more into it :smile:

--
Kasper

**Comment by fbernier** [comment]
The tests are failing [here](https://github.com/rails/rails/blob/33302fdd14d6e82293d1d8ea151af1197e7d74e2/activesupport/test/core_ext/range_ext_test.rb#L109-L114) because of a SystemStackError. After looking into it, I highly suspect it being a ruby bug and I've found [this issue](https://bugs.ruby-lang.org/issues/10847) which sounds like [our problem](https://github.com/ruby/ruby/blob/ca6b174078fa15f33655be704d9409fdbc4f9929/range.c#L1195).

If that's the case, this particular modification will have to wait again I guess.

**Comment by yuki24** [comment]
@fbernier you are right. We can't just replace `#alias_chain_method` with `#prepend` at this point due to that Ruby bug and this only works on Ruby 2.3-head since it's not backported the 2.2-head.

Regarding `#to_json`, you can do something like this:

```ruby
module ActiveSupport
  module CoreExt
    module ToJsonWithActiveSupportEncoder
      def to_json(options = nil)
        if options.is_a?(::JSON::State)
          # Called from JSON.{generate,dump}, forward it to JSON gem's to_json
          super
        else
          # to_json is being invoked directly, use ActiveSupport's encoder
          ActiveSupport::JSON.encode(self, options)
        end
       end
     end
   end
 end
 
[Object, Array, FalseClass, Float, Hash, Integer, NilClass, String, TrueClass, Enumerable].reverse_each do |klass|
  klass.prepend ActiveSupport::CoreExt::ToJsonWithActiveSupportEncoder
end
```

The key here is to use **`#reverse_each`**, not `#reach`.

File: activesupport/test/core_ext/marshal_test.rb (hunk 4 above)
Comment (yuki24) [inline activesupport/test/core_ext/marshal_test.rb]: Do this:

```ruby
assert_equal Marshal.method(:load).super_method.call(dumped), Marshal.load(dumped)
```

**Comment by kirs** [comment]
Thanks @yuki24!

I updated all the code.
I suggest that now we deprecate `alias_method_chain`, migrate to `Module#prepend` everywhere except of `Range` and then patch `Range` as soon as the bug in MRI will be fixed.

@kaspth @rafaelfranca @fbernier what do you think?

File: activesupport/lib/active_support/core_ext/marshal.rb [hunk 5]
```diff
@@ -1,21 +1,17 @@
...
-      if exc.message.match(%r|undefined class/module (.+)|)
-        # try loading the class/module
-        $1.constantize
-        # if it is a IO we need to go back to read the object
-        source.rewind if source.respond_to?(:rewind)
-        retry
-      else
-        raise exc
-      end
+module MarshalWithAutoloading
```
Comment (yuki24) [inline activesupport/lib/active_support/core_ext/marshal.rb]: I think this module should have the `ActiveSupport` namespace.

Comment (rafaelfranca) [inline activesupport/lib/active_support/core_ext/marshal.rb]: :+1:

**Comment by rafaelfranca** [comment]
> I suggest that now we deprecate alias_method_chain, migrate to Module#prepend everywhere except of Range and then patch Range as soon as the bug in MRI will be fixed.

:+1:

**Comment by rafaelfranca** [comment]
But lets add the deprecation in a different PR.

**Comment by kirs** [comment]
@rafaelfranca updated with the namespaced module. I hope it's ready for merge :rocket:

File: activesupport/lib/active_support/core_ext/marshal.rb [hunk 6]
```diff
@@ -1,9 +1,7 @@
-require 'active_support/core_ext/module/aliasing'
-
-module Marshal
-  class << self
-    def load_with_autoloading(source)
-      load_without_autoloading(source)
+module ActiveSupport
+  module MarshalWithAutoloading
```
Comment (rafaelfranca) [inline activesupport/lib/active_support/core_ext/marshal.rb]: `# :nodoc:`

File: activesupport/lib/active_support/core_ext/object/json.rb [hunk 7]
```diff
@@ -26,22 +25,27 @@
...
-      if options.is_a?(::JSON::State)
-        # Called from JSON.{generate,dump}, forward it to JSON gem's to_json
-        self.to_json_without_active_support_encoder(options)
-      else
-        # to_json is being invoked directly, use ActiveSupport's encoder
-        ActiveSupport::JSON.encode(self, options)
+
+module ActiveSupport
+  module CoreExt
+    module ToJsonWithActiveSupportEncoder
```
Comment (rafaelfranca) [inline activesupport/lib/active_support/core_ext/object/json.rb]: `# :nodoc:`

**Comment by rafaelfranca** [comment]
I commented but don't need to change. I'll do.

File: activesupport/lib/active_support/core_ext/object/json.rb (hunk 7 above)
Comment (yuki24) [inline activesupport/lib/active_support/core_ext/object/json.rb]: @kirs @rafaelfranca I should've mentioned this too. Which namespace is more ideal, `ActiveSupport::CoreExt::...` or just `ActiveSupport::...`?

Comment (rafaelfranca) [inline activesupport/lib/active_support/core_ext/object/json.rb]: Just changed to `ActiveSupport::`. 7c7d9dcf5a966a5490fe99230ab9513d73df9630

Comment (yuki24) [inline activesupport/lib/active_support/core_ext/object/json.rb]: :+1:

**Pull request merged**